gossip/tx_scrambler_test.go

```diff
@@ -0,0 +1,645 @@
+		},
+		{
+			hash:   common.Hash{3},
+			sender: common.Address{1},
+			nonce:  3,
+			gas:    1,
+		},
+		{
+			hash:   common.Hash{4},
+			sender: common.Address{1},
+			nonce:  3,
+			gas:    2,
+		},
+		{
+			hash:   common.Hash{5},
+			sender: common.Address{1},
+			nonce:  4,
+		},
+	}
+	shuffleEntries(entries)
+	entries = filterAndOrderTransactions(entries)
+
+	for i := 0; i < len(entries); i++ {
+		for j := i + 1; j < len(entries); j++ {
+			if entries[i].sender == entries[j].sender {
+				if entries[i].nonce > entries[j].nonce {
+					t.Errorf("incorrect nonce order %d must be before %d", entries[j].nonce, entries[i].nonce)
+				}
+				if entries[i].nonce == entries[j].nonce && entries[i].gas < entries[j].gas {
+					t.Errorf("incorrect gas order %d must be before %d", entries[j].gas, entries[i].gas)
+				}
+			}
+		}
+	}
+}
+
+func TestTxScrambler_GetExecutionOrder_RandomInput(t *testing.T) {
+	// this tests these input sizes:
+	// 1, 4, 16, 64, 256, 1024
+	for i := int64(1); i <= 1024; i = i * 4 {
+		input := createRandomScramblerTestInput(i)
+		cpy := deepCopyEntries(input)
+		shuffleEntries(cpy)
+		input = filterAndOrderTransactions(input)
+		cpy = filterAndOrderTransactions(input)
+		if !reflect.DeepEqual(input, cpy) {
+			t.Error("slices have different order - algorithm is not deterministic")
+		}
+	}
+
+}
+
+func BenchmarkTxScrambler(b *testing.B) {
+	for size := int64(10); size < 100_000; size *= 10 {
+		b.Run(fmt.Sprintf("size_%d", size), func(b *testing.B) {
+			for i := 1; i <= b.N; i++ {
+				b.StopTimer()
+				entries := createRandomScramblerTestInput(size)
+				b.StartTimer()
+				filterAndOrderTransactions(entries)
+			}
+		})
+	}
+}
+
+func createRandomScramblerTestInput(size int64) []*scramblerEntry {
+	var entries []*scramblerEntry
+	for i := int64(0); i < size; i++ {
+		entries = append(entries, &scramblerEntry{
+			hash:   common.Hash{byte(rand.Intn(1000 - 1))},
+			sender: common.Address{byte(rand.Intn(100 - 1))},
+			nonce:  uint64(rand.Intn(10 - 1)),
+			gas:    uint64(rand.Intn(10 - 1)),
+		})
+	}
+
+	return entries
+}
+
+// shuffleEntries shuffles given entries randomly.
+func shuffleEntries(entries []*scramblerEntry) {
+	rand.Shuffle(len(entries), func(i, j int) {
+		entries[i], entries[j] = entries[j], entries[i]
+	})
+}
+
+// checkDuplicateHashes checks hash of each entry and fails test if duplicate hash is found.
+func checkDuplicateHashes(t *testing.T, entries []*scramblerEntry) {
+	seenHashes := make(map[common.Hash]struct{})
+	for _, entry := range entries {
+		if _, found := seenHashes[entry.hash]; found {
+			t.Fatalf("found duplicate hash in entries: %s", entry.hash)
+		}
+		seenHashes[entry.hash] = struct{}{}
+	}
+}
+
+func createRandomSalt() [32]byte {
+	var salt = [32]byte{}
+	for i := 0; i < 32; i++ {
+		salt[i] = byte(rand.Intn(math.MaxUint8))
+	}
+	return salt
+}
+
+// deepCopyEntries returns deep copy of entries.
+func deepCopyEntries(entries []*scramblerEntry) []*scramblerEntry {
```

Comment:
nit: I think nowhere in these tests you actually need a deep copy and a deep equality since all entries are all constant. Shallow copies of slices and comparisons would do the job.